clamav/cnudie.py

```diff
@@ -21,22 +25,103 @@
+        resource_node=resource_node,
+        types=[dso.model.Datatype.MALWARE],
+    )
+
+    sorted_findings = sorted(
+        previous_findings,
+        key=lambda f: f.meta.creation_date.astimezone(),
+        reverse=True,
+    )
+
+    if sorted_findings:
+        latest_finding = sorted_findings[0]
+    else:
+        return False
+
+    if isinstance(latest_finding.data, dict):
+        # In this case the previous scan we have available predates the current dataclasses.
+        # Returning False here results in a rescan which will put a new scan result into
+        # the database if there are any findings.
+        # Also, as an additional safeguard against future errors, we check that no such
+        # supposedly structurally outdated finding was put into the database after this
+        # change was added.
+        cutoff_date = datetime.datetime(year=2023, month=4, day=1)
+        if latest_finding.meta.creation_date > cutoff_date:
+            raise RuntimeError(
+                'Encountered a malware scan result with outdated schema that was created '
+                'after introduction of the new schema.'
+            )
+        return False
+
+    finding_clamav_metadata = latest_finding.data.metadata
+
+    if (
+        skip_if_virus_definition_unchanged
+        and finding_clamav_metadata.signature_version == current_version_info.signature_version
+    ):
+        # As of time of writing we check twice a day for signature definition updates and ClamAV
+        # updates the database "usually [...] once or twice per day"
+        # (https://docs.clamav.net/faq/faq-cvd.html). As a result this check is mostly
+        # superfluous since we run most scan-jobs daily, however it could prove useful in case
+        # of future changes/disruptions to the update schedule (either on our side or theirs).
+        return True
+
+    if (
+        current_version_info.signature_date
+        - finding_clamav_metadata.virus_definition_timestamp
+    ).days < virus_db_max_age_days:
+        return True
+
+    return False
+
+
 def scan_resources(
     resource_nodes: typing.Iterable[cnudie.iter.ResourceNode],
+    clamav_version_info: clamav.model.ClamAVVersionInfo,
+    virus_db_max_age_days: int,
+    skip_if_virus_definition_unchanged: bool,
     oci_client: oci.client.Client,
     clamav_client: clamav.client.ClamAVClient,
-    clamav_version_info: clamav.model.ClamAVVersionInfo,
+    delivery_client: delivery.client.DeliveryServiceClient = None,
     s3_client=None,
     max_workers:int = 16,
 ) -> typing.Generator[clamav.model.ClamAVResourceScanResult, None, None]:
-    executor = concurrent.futures.ThreadPoolExecutor(max_workers=max_workers)
 
     def scan_resource(
         resource_node: cnudie.iter.ResourceNode,
-    ):
+    ) -> clamav.model.ClamAVResourceScanResult:
         component = resource_node.component
         resource = resource_node.resource
 
+        if _should_skip_scan(
```

Comment:
I think this should be split + renamed, still, like so:

`_may_reuse_existing_scan_result`: check whether we are "allowed" to reuse existing scan-result (we might have been called w/ a "force-scan" flag. however: why do we even want to not try to reuse existing scans?

```scan_result = _find_scan_result(
 max_age, # hardcode to ignore if version matches
 version, # make version optional
 <ocm-coordinates>, 
 ..
)
```